passes/opt/opt_hier.cc

```diff
@@ -0,0 +1,470 @@
+					constant_inputs.erase(port_bit);
+				}
+			}
+		}
+	}
+
+	void refine_tie_togethers(const dict<SigBit, SigBit> &inputs) {
+		std::vector<SigSpec> new_tie_togethers;
+
+		for (auto &class_ : tie_together_inputs) {
+			dict<SigBit, SigSpec> new_classes;
+
+			for (auto bit : class_) {
+				SigBit connected_bit = inputs.count(bit) ? inputs.at(bit) : RTLIL::Sx;
+				new_classes[connected_bit].append(bit);
+			}
+
+			for (auto [key, new_class] : new_classes) {
+				if (new_class.size() > 1)
+					new_tie_togethers.push_back(new_class);
+			}
+		}
+
+		new_tie_togethers.swap(tie_together_inputs);
+	}
+
+	// inspect the given instantiation and refine usage data accordingly
+	void refine(Cell *instance, ModuleIndex &index) {
+		dict<SigBit, SigBit> inputs;
+
+		for (auto &[port_name, value] : instance->connections_) {
+			Wire *port = module->wire(port_name);
+			if (!port || (!port->port_input && !port->port_output) || port->width != value.size()) {
+				log_error("Port %s connected on instance %s not found in module %s"
+						  " or width is not matching\n",
+						  log_id(port_name), log_id(instance), log_id(module));
+			}
+
+			if (port->port_input && port->port_output) {
+				// ignore bidirectional: hard to come up with sound handling
+				continue;
+			}
+
+			if (port->port_output) {
+				refine_used_outputs(port, value, index);
+			} else {
+				refine_input_constants(port, value);
+				for (int i = 0; i < port->width; i++)
+					inputs[SigBit(port, i)] = value[i];
+			}
+		}
+
+		refine_tie_togethers(inputs);
+	}
+
+	bool apply_changes() {
+		bool did_something = false;
+
+		if (module->get_blackbox_attribute()) {
+			// no propagating into blackboxes
+			return false;
+		}
+
+		// Disconnect unused outputs
+		for (auto &pair : module->connections_) {
+			for (int i = 0; i < pair.first.size(); i++) {
+				// If an output is constant there's no benefit to disconnecting
+				// so consider it "used"
+				if (pair.first[i].wire
+						&& pair.first[i].wire->port_output
+						&& !pair.second[i].wire)
+					used_outputs.add(pair.first[i]);
+			}
+		}
+
+		SigSpec disconnect_outputs;
+		for (auto bit : all_outputs) {
+			if (!used_outputs.check(bit))
+				disconnect_outputs.append(bit);
+		}
+
+		dict<SigBit, SigBit> replacement_map;
+		for (auto chunk : disconnect_outputs.chunks()) {
+			Wire *repl_wire = module->addWire(module->uniquify(std::string("$") + chunk.wire->name.str()), chunk.size());
+			for (int i = 0; i < repl_wire->width; i++)
+				replacement_map[SigSpec(chunk)[i]] = SigBit(repl_wire, i);
+		}
+		auto disconnect_rewrite = [&](SigSpec &signal) {
+			signal.replace(replacement_map);
+		};
+		module->rewrite_sigspecs(disconnect_rewrite);
+		for (auto chunk : disconnect_outputs.chunks()) {
+			log("Disconnected unused output terminal '%s' in module '%s'\n", log_signal(chunk), log_id(module));
+			did_something = true;
+			module->connect(chunk, SigSpec(RTLIL::Sx, chunk.size()));
+		}
+
+		// Connect constant inputs
```

Comment:
generally having a long sequence of non-trivial operations that can be grouped with comments into distinct sections suggests they could be separated into functions. Consider breaking these up